### PR TITLE
Failed the links with 'en-us' locale

### PR DESCRIPTION
**Pull request by sima-zhu**
I have tested over the scripts in andriod pipeline. https://dev.azure.com/azure-sdk/public/_build/results?buildId=493840&view=logs&jobId=b70e5e73-bbb6-5567-0939-8415943fadb9&j=b70e5e73-bbb6-5567-0939-8415943fadb9&t=e75da802-0139-5977-bd1e-91175bec9059

### CODE REVIEW

File: eng/common/scripts/Verify-Links.ps1 [hunk 1]
```diff
@@ -6,18 +6,20 @@ param (
...
   [array] $errorStatusCodes = @(400, 401, 404, 11001, 11004),
-  # flag to allow resolving relative paths or not
-  [bool] $resolveRelativeLinks = $true
+  # flag to allow checking against azure sdk link guidance.
+  [bool] $linkGuidance = $true
 )
 
 $ProgressPreference = "SilentlyContinue"; # Disable invoke-webrequest progress dialog
+# list of locale keywords
+[array] $locales = @("/en-us/")
```
Comment (weshaggard) [inline eng/common/scripts/Verify-Links.ps1:22]: I wouldn't make this a list I would just make it a regex.

File: eng/common/scripts/Verify-Links.ps1 [hunk 2]
```diff
@@ -65,11 +67,11 @@ function ResolveUri ([System.Uri]$referralUri, [string]$link)
   }
 
   $linkUri = [System.Uri]$link;
-  if($resolveRelativeLinks){
+  if($linkGuidance){
```
Comment (weshaggard) [inline eng/common/scripts/Verify-Links.ps1:70]:
```suggestion
  if($linkGuidance) {
```

nit: space before brace

File: eng/common/scripts/Verify-Links.ps1 [hunk 3]
```diff
@@ -117,13 +119,26 @@ function CheckLink ([System.Uri]$linkUri)
   if ($checkedLinks.ContainsKey($linkUri)) { return }
 
   Write-Verbose "Checking link $linkUri..."
+
+  # Check if link uri includes locale info.
+  if ($linkGuidance) {
+    foreach ($locale in $locales) {
```
Comment (weshaggard) [inline eng/common/scripts/Verify-Links.ps1:125]: The foreach won't be needed when you make the locales a regex.

File: eng/common/scripts/Verify-Links.ps1 [hunk 4]
```diff
@@ -117,13 +119,26 @@ function CheckLink ([System.Uri]$linkUri)
   if ($checkedLinks.ContainsKey($linkUri)) { return }
 
   Write-Verbose "Checking link $linkUri..."
+
+  # Check if link uri includes locale info.
+  if ($linkGuidance) {
```
Comment (weshaggard) [inline eng/common/scripts/Verify-Links.ps1:124]: Lets move this to the bottom of this function and we should make sure we don't add this to badLinks more then once

File: eng/common/scripts/Verify-Links.ps1 [hunk 5]
```diff
@@ -6,18 +6,20 @@ param (
...
   # recusiving check links for all links verified that begin with this baseUrl, defaults to the folder the url is contained in
   [string] $baseUrl = "",
   # path to the root of the site for resolving rooted relative links, defaults to host root for http and file directory for local files
   [string] $rootUrl = "",
   # list of http status codes count as broken links. Defaults to 400, 401, 404, SocketError.HostNotFound = 11001, SocketError.NoData = 11004
   [array] $errorStatusCodes = @(400, 401, 404, 11001, 11004),
-  # flag to allow resolving relative paths or not
-  [bool] $resolveRelativeLinks = $true
+  # flag to allow checking against azure sdk link guidance.
+  [bool] $linkGuidance = $true
```
Comment (weshaggard) [inline eng/common/scripts/Verify-Links.ps1:17]: I'd rename this to "checkLinkGuidance" and set it to false by default.

File: eng/common/scripts/Verify-Links.ps1 [hunk 6]
```diff
@@ -6,18 +6,20 @@ param (
   # switch that will enable devops specific logging for warnings
   [switch] $devOpsLogging = $false,
   # check the links recurisvely based on recursivePattern
-  [switch] $recursive = $true,
+  [switch] $recursive,
```
Comment (weshaggard) [inline eng/common/scripts/Verify-Links.ps1:9]: Why did you change the default here?

File: eng/common/pipelines/templates/steps/verify-links.yml [hunk 7]
```diff
@@ -11,4 +11,4 @@ steps:
       workingDirectory: $(Build.SourcesDirectory)/${{ parameters.Directory }}
       filePath: eng/common/scripts/Verify-Links.ps1 
       arguments: >
-        -urls $(dir -r -i *.md) -rootUrl "file://$(Build.SourcesDirectory)/${{ parameters.Directory }}" -recursive:$false -ignoreLinksFile ${{ parameters.IgnoreLinksFile }}
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/verify-links.yml:14]: Can you split this into multiple lines? With `>` already there you can just split.

```
-urls $(dir -r -i *.md) 
-rootUrl "file://$(Build.SourcesDirectory)/${{ parameters.Directory }}" 
-recursive:$false 
-ignoreLinksFile ${{ parameters.IgnoreLinksFile }}
```

File: eng/common/scripts/Verify-Links.ps1 (hunk 1 above)
Comment (sima-zhu) [reply eng/common/scripts/Verify-Links.ps1:22]: We might add more locale strings in future. Make it regex gives us limited room to extend, but I am flexible on the data structure here.

File: eng/common/scripts/Verify-Links.ps1 (hunk 2 above)
Comment (sima-zhu) [reply eng/common/scripts/Verify-Links.ps1:70]: Will make the change together with renaming.

File: eng/common/scripts/Verify-Links.ps1 (hunk 6 above)
Comment (sima-zhu) [reply eng/common/scripts/Verify-Links.ps1:9]: The IDE reminds me switch defaults the bool field to true.
I tested it a bit. It is not default to true. Will revert the change back.

File: eng/common/scripts/Verify-Links.ps1 (hunk 5 above)
Comment (sima-zhu) [reply eng/common/scripts/Verify-Links.ps1:17]: Relative link is enabled right now. Will rename this, to make the feature backward compatible, will enable relative link and disable the check on locale

File: eng/common/scripts/Verify-Links.ps1 (hunk 4 above)
Comment (sima-zhu) [reply eng/common/scripts/Verify-Links.ps1:124]: Will move to bottom. If we have a broken link with locale, then we will print warning messages, but only display one bad link.

File: eng/common/scripts/Verify-Links.ps1 [hunk 8]
```diff
@@ -115,15 +117,19 @@ function ParseLinks([string]$baseUri, [string]$htmlContent)
 function CheckLink ([System.Uri]$linkUri)
 {
   if ($checkedLinks.ContainsKey($linkUri)) { return }
+  $badLinkEnabled = $false;
```
Comment (weshaggard) [inline eng/common/scripts/Verify-Links.ps1:120]: I've been wanting to update this function to return true if the link is valid and false if the link is not valid. So can you instead rename this to `$linkValid = $true` and then set it to false anytime we detect it is bad? Then update the $checkedLinks dictionary to store the value and return it at the top of the function and return `$linkValid` at the end of this function.

File: eng/common/scripts/Verify-Links.ps1 (hunk 1 above)
Comment (weshaggard) [reply eng/common/scripts/Verify-Links.ps1]: Adding more locales can be added to the regex but if that doesn't end up working for a future scenario then we can make adjustments then. For now I'd like to keep it as simple as possible.

File: eng/common/scripts/Verify-Links.ps1 [hunk 9]
```diff
@@ -115,15 +117,19 @@ function ParseLinks([string]$baseUri, [string]$htmlContent)
...
-  Write-Verbose "Checking link $linkUri..."
   if ($linkUri.IsFile) {
     if (!(Test-Path $linkUri.LocalPath)) {
       LogWarning "Link to file does not exist $($linkUri.LocalPath)"
       $script:badLinks += $linkUri
+      $badLinkEnabled = $true
     }
   }
   else {
+
```
Comment (weshaggard) [inline eng/common/scripts/Verify-Links.ps1:132]: nit: Lets remove this empty line so that the try block is at the top of the block.

File: eng/common/scripts/Verify-Links.ps1 (hunk 8 above)
Comment (sima-zhu) [reply eng/common/scripts/Verify-Links.ps1:120]: Fixed as you suggested, much better for the entire structure.
$checkedLinks currently stores like : {<link>: <valid or not>}

File: eng/common/scripts/Verify-Links.ps1 [hunk 10]
```diff
@@ -115,12 +117,14 @@ function ParseLinks([string]$baseUri, [string]$htmlContent)
 function CheckLink ([System.Uri]$linkUri)
 {
   if ($checkedLinks.ContainsKey($linkUri)) { return }
+  $checkedLinks[$linkUri] = $true;
```
Comment (weshaggard) [inline eng/common/scripts/Verify-Links.ps1:120]: We should not directly access the hashtable through the this function as it might become a threading issue if we ever try to run this in parallel. Instead please use a local `$linkValid` variable instead and only read the hashtable at the top and set it at the bottom of this function.

File: eng/common/scripts/Verify-Links.ps1 [hunk 11]
```diff
@@ -115,12 +117,14 @@ function ParseLinks([string]$baseUri, [string]$htmlContent)
 function CheckLink ([System.Uri]$linkUri)
 {
   if ($checkedLinks.ContainsKey($linkUri)) { return }
```
Comment (weshaggard) [inline eng/common/scripts/Verify-Links.ps1:119]: This should now return `$checkedLinks[$linkUri]`.

File: eng/common/scripts/Verify-Links.ps1 [hunk 12]
```diff
@@ -165,7 +170,14 @@ function CheckLink ([System.Uri]$linkUri)
...
-  $checkedLinks[$linkUri] = $true;
+  
+  # Check if link uri includes locale info.
+  if ($checkLinkGuidance -and ($linkUri -match $locale)) {
+    LogWarning "DO NOT include locale 'en-us' information in links: $linkUri."
+    if ($checkedLinks[$linkUri]) {
+      $script:badLinks += $linkUri
+    }
+  }
 }
```
Comment (weshaggard) [inline eng/common/scripts/Verify-Links.ps1:181]: We should return `$linkValid` once you setup the local value.

File: eng/common/scripts/Verify-Links.ps1 [hunk 13]
```diff
@@ -65,11 +67,11 @@ function ResolveUri ([System.Uri]$referralUri, [string]$link)
   }
 
   $linkUri = [System.Uri]$link;
-  if($resolveRelativeLinks){
+  if(-Not $checkLinkGuidance) {
```
Comment (weshaggard) [inline eng/common/scripts/Verify-Links.ps1:70]: It is probably worth adding a comment here on why resolving relative links is gated by the checkLinkGuidance variable.

File: eng/common/scripts/Verify-Links.ps1 (hunk 11 above)
Comment (sima-zhu) [reply eng/common/scripts/Verify-Links.ps1:119]: The CheckLink function does not return anything. Do we want to change it to return boolean?

**Comment by sima-zhu** [comment]
https://dev.azure.com/azure-sdk/public/_build/results?buildId=496990&view=logs&jobId=b70e5e73-bbb6-5567-0939-8415943fadb9&j=b70e5e73-bbb6-5567-0939-8415943fadb9&t=e75da802-0139-5977-bd1e-91175bec9059

File: eng/common/scripts/Verify-Links.ps1 [hunk 14]
```diff
@@ -165,7 +174,17 @@ function CheckLink ([System.Uri]$linkUri)
       }
     }
   }
-  $checkedLinks[$linkUri] = $true;
+  
+  # Check if link uri includes locale info.
+  if ($checkLinkGuidance -and ($linkUri -match $locale)) {
+    LogWarning "DO NOT include locale $locale information in links: $linkUri."
+    if ($linkValid) {
+      $script:badLinks += $linkUri
```
Comment (weshaggard) [inline eng/common/scripts/Verify-Links.ps1:182]: Lets please move all the additions of badLinks from this function as well and do it where we call CheckLink

File: eng/common/scripts/Verify-Links.ps1 [hunk 15]
```diff
@@ -265,6 +261,9 @@ while ($pageUrisToCheck.Count -ne 0)
   
   foreach ($linkUri in $linkUris) {
     $isLinkValid = CheckLink $linkUri
+    if (!linkUri) {
```
Comment (weshaggard) [inline eng/common/scripts/Verify-Links.ps1:264]:
```suggestion
    if (!isLinkValid) {
```

File: eng/common/scripts/Verify-Links.ps1 [hunk 16]
```diff
@@ -65,11 +67,14 @@ function ResolveUri ([System.Uri]$referralUri, [string]$link)
   }
 
   $linkUri = [System.Uri]$link;
-  if($resolveRelativeLinks){
+  # Relative link check is gated by flag $checkLinkGuidance. 
+  # Allows to use relative link by default for backward compatibility.
+  # TODO: Need to disallow relative links in accordance with new sdk guidance.
```
Comment (weshaggard) [inline eng/common/scripts/Verify-Links.ps1:72]:
```suggestion
  # Our link guidelines do not allow relative links so only resolve them when we are not
  # validating links against our link guidelines (i.e. !$checkLinkGuideance)
```

File: eng/common/scripts/Verify-Links.ps1 [hunk 17]
```diff
@@ -114,13 +119,15 @@ function ParseLinks([string]$baseUri, [string]$htmlContent)
 
 function CheckLink ([System.Uri]$linkUri)
 {
-  if ($checkedLinks.ContainsKey($linkUri)) { return }
+  if ($checkedLinks.ContainsKey($linkUri)) { return $checkedLinks[$linkUri]}
```
Comment (weshaggard) [inline eng/common/scripts/Verify-Links.ps1:122]:
```suggestion
  if ($checkedLinks.ContainsKey($linkUri)) { 
    return $checkedLinks[$linkUri] 
  }
```

File: eng/common/scripts/Verify-Links.ps1 [hunk 18]
```diff
@@ -165,7 +172,16 @@ function CheckLink ([System.Uri]$linkUri)
...
     }
   }
-  $checkedLinks[$linkUri] = $true;
+  
+  # Check if link uri includes locale info.
+  if ($checkLinkGuidance -and ($linkUri -match $locale)) {
+    LogWarning "DO NOT include locale $locale information in links: $linkUri."
+    if ($linkValid) {
+      $linkValid = $false
+    }
```
Comment (weshaggard) [inline eng/common/scripts/Verify-Links.ps1:181]:
```suggestion
    $linkValid = $false
```

File: eng/common/pipelines/templates/steps/verify-links.yml [hunk 19]
```diff
@@ -11,4 +11,7 @@ steps:
...
+        -urls $(dir -r -i *.md) 
+        -rootUrl "file://$(Build.SourcesDirectory)/${{ parameters.Directory }}" 
+        -recursive $false
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/verify-links.yml:16]: Please double check this I believe it needs to be passed like `-recursive:$false`
